CASMUSER-3182: Backport K3s support

alanm-hpe opened the pull request:
## Summary and Scope

Backport K3s support to `release/uan-2.5`

Also introduce a fix for airgapped systems.
* use haproxy image from nexus
* use metallb controller+speaker image from nexus
* disable unused k3s components and use nexus pause image

## Issues and Related PRs 

* Resolves [CASMUSER-3812](https://jira-pro.its.hpecorp.net:8443/browse/CASMUSER-3182)
* Requires https://github.com/Cray-HPE/uan/pull/69

## Testing

### Tested on:

  * `lemondrop`
  * `groot`

### Test description:

Installed a new UAN product with updated CFS. Used a custom built image due to MQTT DVS work on `groot`.

Full k3s environment was created and executed the K3s Test plan to run podman containers as UAI login shells.

Was not able to provide load balancer IP addresses to metallb as `groot` is an airgapped system.
```
uan01:~ # kubectl get services -n haproxy-uai
NAME          TYPE           CLUSTER-IP    EXTERNAL-IP   PORT(S)        AGE
haproxy-uai   LoadBalancer   10.43.94.68   <pending>     22:31024/TCP   48m
uan01:~ # ssh alanm@10.43.94.68
Password:
sh-4.4$ sinfo
PARTITION AVAIL  TIMELIMIT  NODES  STATE NODELIST
workq*       up   infinite      2  down* nid[000001,000004]
workq*       up   infinite      2   idle nid[000002-000003]
```

## Risks and Mitigations

K3s is an isolated playbook so it needs to be introduced to a UAN CFS configuration in order to be activated.


## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [ ] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

darrenn-jackson commented on docs/portal/developer-portal/test-plan/K3s Test Plan.md:
```diff
@@ -0,0 +1,195 @@
+This procedure assumes familiarity with the commands `cray cfs`, `cray bos` and, optionally, access to `sat bootprep` automation files to assist in reconfiguration of the UAN.
+
+## Configuration  
+To execute this test plan, first identify a UAN that has been fully booted and configured with COS. This means the UAN is configured with the CAN or CHN, and is able to authenticate non-root users.
```
```suggestion
1. Identify a UAN that has been fully booted and configured with COS. 

  The UAN chosen must be configured with the CAN or CHN, and be able to authenticate non-root users.
  
```

darrenn-jackson commented on docs/portal/developer-portal/test-plan/K3s Test Plan.md:
````diff
@@ -0,0 +1,195 @@
+default via x.x.x.x dev can0
+```
+
+1. Verify non-root SSH access to the UAN (make sure to use chn or can based on the default route):
````
```suggestion
1. Verify non-root SSH access to the UAN.  Use `chn` or `can` depending on the default route):
```

darrenn-jackson commented on docs/portal/developer-portal/test-plan/K3s Test Plan.md:
````diff
@@ -0,0 +1,195 @@
+alanm@uan01:~
+```
+
+1. Once a target UAN has been identified to test, checkout the UAN CFS repository from VCS, and create a new branch. In the following example, the currently booted UAN was configured with `integration-2.6.0`, so that was used as the starting point. Use the correct branch for the UAN being tested.
````
```suggestion
1. Checkout the UAN CFS repository from VCS and create a new branch. 

  In the following example, the currently booted UAN was configured with `integration-2.6.0`, so that was used as the starting point. Use the correct branch for the UAN being tested.
```

darrenn-jackson commented on docs/portal/developer-portal/test-plan/K3s Test Plan.md:
````diff
@@ -0,0 +1,195 @@
+Switched to branch 'k3s'
+Your branch is up to date with 'origin/integration-2.6.0'.
+```
+1. Complete the [MetalLB section](../advanced/Enabling_K3s.md#metallb) of the Enabling K3s guide. Once completed, modify `vars/uan_helm.yml` (IP addresses masked with `x.x.x.x`) and commit the change.
````
```suggestion
1. Complete the [MetalLB section](../advanced/Enabling_K3s.md#metallb) of the Enabling K3s guide. 

1. Modify `vars/uan_helm.yml` (IP addresses masked with `x.x.x.x`) and commit the change.
```

darrenn-jackson commented on docs/portal/developer-portal/test-plan/K3s Test Plan.md:
````diff
@@ -0,0 +1,195 @@
+[k3s 1dde4ab] Add MetalLB IP Address range
+ 1 file changed, 2 insertions(+), 2 deletions()
+```
+1. Complete the [HAProxy section](../advanced/Enabling_K3s.md#haproxy-configuration) of the Enabling K3s guide. The following changes to `vars/uan_helm.yml` would be suitable changes for validation (make sure to change the DNS name for the UAN selected for this test):
````
```suggestion
1. Complete the [HAProxy section](../advanced/Enabling_K3s.md#haproxy-configuration) of the Enabling K3s guide. Replace `uan01 uan01.can.lemondrop.hpc.amslabs.hpecorp.net:9000` with the DNS of the UAN chosen for this test.

The following changes to `vars/uan_helm.yml` would be suitable changes for validation:
```

darrenn-jackson commented on docs/portal/developer-portal/test-plan/K3s Test Plan.md:
````diff
@@ -0,0 +1,195 @@
+[k3s 1b54db9] Add HAProxy configuration
+ 1 file changed, 19 insertions(+)
+```
+1. Complete the  [SSHD section](../advanced/Enabling_K3s.md#sshd-configuration) of the Enabling K3s guide. These changes to `vars/uan_sshd.yml` may be used as a valid test of a podman configuration:
````
```suggestion
1. Complete the  [SSHD section](../advanced/Enabling_K3s.md#sshd-configuration) of the Enabling K3s guide. 

  These changes to `vars/uan_sshd.yml` may be used as a valid test of a podman configuration:
```

darrenn-jackson commented on docs/portal/developer-portal/test-plan/K3s Test Plan.md:
````diff
@@ -0,0 +1,195 @@
+if ! grep -sq $TEST_USER /etc/subuid; then echo $TEST_USER:200000000:$(id -u $TEST_USER) >> /etc/subuid; fi
+if ! grep -sq $TEST_USER /etc/subgid; then echo $TEST_USER:200000000:$(id -u $TEST_USER) >> /etc/subgid; fi
+```
+1. Create and upload a podman container to the CSM registry. The following example uses the default container image from the User Access Service and should be run from an NCN:
````
```suggestion
1. Create and upload a podman container to the CSM registry. 

The following example uses the default container image from the User Access Service and should be run from an NCN:
```

darrenn-jackson commented on docs/portal/developer-portal/test-plan/K3s Test Plan.md:
````diff
@@ -0,0 +1,195 @@
+
+vim $CFS_CONFIG.json
+```
+1. The modified CFS configuration should include the following new playbook after the standard UAN `site.yaml` playbook:
````
```suggestion
The modified CFS configuration should include the following new playbook after the standard UAN `site.yaml` playbook:
```

darrenn-jackson commented on docs/portal/developer-portal/advanced/Enabling_K3s.md:
```diff
@@ -0,0 +1,323 @@
+
+To validate the K3s cluster once deployed, see the [Validation Checks](#validation-checks) section of this document for details.
+
+## Configuring K3s, MetalLB, HAProxy, and SSHD for use with Podman
```
Since this is a 2nd level heading just like the Prerequisites, I assume that this is the actual procedure for enabling K3s?

alanm-hpe replied:
Yes

darrenn-jackson commented on docs/portal/developer-portal/advanced/Enabling_K3s.md:
```diff
@@ -0,0 +1,323 @@
+
+## Configuring K3s, MetalLB, HAProxy, and SSHD for use with Podman
+
+Each of the sections below describe how the various comonents deployed to K3s and the UANs may be configured to enable users to SSH to rootless podman containers. As there is no one configuration to fit any one use case, read and understand each section to modify the configuration as needed. Once each section has been completed, see [Deploy K3s to the UAN](#deploy-k3s-to-the-uan).
```
"comonents" should be "components"

darrenn-jackson replied:
```suggestion
Each of the sections below describe how the various components deployed to K3s and the UANs may be configured to enable users to SSH to rootless podman containers. As there is no one configuration to fit any one use case, read and understand each section to modify the configuration as needed. Once each section has been completed, see [Deploy K3s to the UAN](#deploy-k3s-to-the-uan).
```

darrenn-jackson reviewed (changes requested):
Several simple formatting changes for readability/skimmability. One spelling error and a couple of questions.

alanm-hpe commented on docs/portal/developer-portal/uan_admin.ditamap:
```diff
@@ -27,6 +27,7 @@
         <topicref href="advanced/Repurposing_Compute_as_UAN.md" format="markdown"/>
         <topicref href="advanced/Enabling_CAN_CHN.md" format="markdown"/>
         <topicref href="advanced/SLES_Image.md" format="markdown"/>
+        <topicref href="advanced/Enabling_K3s.md" format="markdown"/>
```
This is a PR to `uan-product-stream` for `release/uan-2.5` to make sure we have the docs in this release. Separately, I will make the change to `docs-uan` for 2.5.

darrenn-jackson reviewed (changes requested):
one typo to fix

alanm-hpe commented:
UAN install log
```
ncn-m001:/mnt/developer/alanm/uan-2.5.9-7-g32bfb7f # ./install.sh
+ trap notify ERR
++ dirname ./install.sh
+ ROOTDIR=.
++ craysys metadata get system-name
+ SYS_NAME=groot
++ craysys metadata get site-domain
+ SITE_DOMAIN=hpc.amslabs.hpecorp.net
+ API_GW=https://api.nmnlb.groot.hpc.amslabs.hpecorp.net
+ source ./lib/install.sh
++ : https://packages.local
++ : registry.local
++ export NEXUS_URL
++ : .
++ podman_run_flags=(--network host)
++ declare -a podman_run_flags
++ [[ no == \y\e\s ]]
++ requires curl find podman realpath
++ [[ 4 -gt 0 ]]
++ command -v curl
++ shift
++ [[ 3 -gt 0 ]]
++ command -v find
++ shift
++ [[ 2 -gt 0 ]]
++ command -v podman
++ shift
++ [[ 1 -gt 0 ]]
++ command -v realpath
++ shift
++ [[ 0 -gt 0 ]]
++ vendor_images=()
++ declare -a vendor_images
+ source ./vars.sh
++ UAN_PRODUCT_VERSION=2.5.9-7-g32bfb7f
++ UAN_CONFIG_VERSION=1.9.11
++ PRODUCT_CATALOG_UPDATE_VERSION=1.3.1
++ UAN_IMAGE_VERSION=0.1.0
++ UAN_IMAGE_NAME=cray-application-sles15sp3.x86_64-0.1.0
++ UAN_KERNEL_VERSION=5.3.18-150300.59.87-default
+++ realpath .
++ basename /mnt/developer/alanm/uan-2.5.9-7-g32bfb7f
+ : uan-2.5.9-7-g32bfb7f
+ mkdir -p ./build/manifests
+ manifestgen -i ./manifests/uan.yaml -o ./build/manifests/uan.yaml
+ load-install-deps
+ [[ -f ./vendor/cray-nexus-setup.tar ]]
+ [[ -v CRAY_NEXUS_SETUP_IMAGE ]]
++ load-vendor-image ./vendor/cray-nexus-setup.tar
++ set -o pipefail
++ podman load -q -i ./vendor/cray-nexus-setup.tar
++ sed -e 's/^.*: //'
+ CRAY_NEXUS_SETUP_IMAGE=docker.io/library/cray-nexus-setup:uan-2.5.9-7-g32bfb7f
+ vendor_images+=("$CRAY_NEXUS_SETUP_IMAGE")
+ [[ -f ./vendor/skopeo.tar ]]
+ [[ -v SKOPEO_IMAGE ]]
++ load-vendor-image ./vendor/skopeo.tar
++ set -o pipefail
++ podman load -q -i ./vendor/skopeo.tar
++ sed -e 's/^.*: //'
+ SKOPEO_IMAGE=docker.io/library/skopeo:uan-2.5.9-7-g32bfb7f
+ vendor_images+=("$SKOPEO_IMAGE")
+ nexus-setup blobstores ./nexus-blobstores.yaml
+ nexus-setdefault-credential
+ [[ -v NEXUS_PASSWORD ]]
+ nexus-get-credential
+ requires base64 kubectl
+ [[ 2 -gt 0 ]]
+ command -v base64
+ shift
+ [[ 1 -gt 0 ]]
+ command -v kubectl
+ shift
+ [[ 0 -gt 0 ]]
+ [[ 0 -gt 0 ]]
+ set -- -n nexus nexus-admin-credential
+ kubectl get secret -n nexus nexus-admin-credential
++ base64 -d
++ kubectl get secret -n nexus nexus-admin-credential -o 'jsonpath={.data.username}'
+ export NEXUS_USERNAME=admin
+ NEXUS_USERNAME=admin
++ kubectl get secret -n nexus nexus-admin-credential -o 'jsonpath={.data.password}'
++ base64 -d
+ export NEXUS_PASSWORD=EcvveawHHqtQPkAOEapCdRiVDYUrmz5VIr9FE6-vanU=
+ NEXUS_PASSWORD=EcvveawHHqtQPkAOEapCdRiVDYUrmz5VIr9FE6-vanU=
+ return 0
++ realpath ./nexus-blobstores.yaml
+ podman run --rm --network host -v /mnt/developer/alanm/uan-2.5.9-7-g32bfb7f/nexus-blobstores.yaml:/config.yaml:ro -e NEXUS_URL -e NEXUS_USERNAME -e NEXUS_PASSWORD docker.io/library/cray-nexus-setup:uan-2.5.9-7-g32bfb7f nexus-blobstores-create /config.yaml
Creating file blobstore: uan... 400 FAIL
Updating file blobstore: uan... 204 OK updated
+ nexus-setup repositories ./nexus-repositories.yaml
+ nexus-setdefault-credential
+ [[ -v NEXUS_PASSWORD ]]
+ [[ -n EcvveawHHqtQPkAOEapCdRiVDYUrmz5VIr9FE6-vanU= ]]
+ return 0
++ realpath ./nexus-repositories.yaml
+ podman run --rm --network host -v /mnt/developer/alanm/uan-2.5.9-7-g32bfb7f/nexus-repositories.yaml:/config.yaml:ro -e NEXUS_URL -e NEXUS_USERNAME -e NEXUS_PASSWORD docker.io/library/cray-nexus-setup:uan-2.5.9-7-g32bfb7f nexus-repositories-create /config.yaml
Looking for raw/hosted repository: uan-2.5.9-sle-15sp4... Found...Updating raw/hosted repository: uan-2.5.9-sle-15sp4...204 OK updated
Looking for raw/group repository: uan-2.5-sle-15sp4... Found...Updating raw/group repository: uan-2.5-sle-15sp4...204 OK updated
Looking for raw/hosted repository: uan-2.5.9-sle-15sp3... Found...Updating raw/hosted repository: uan-2.5.9-sle-15sp3...204 OK updated
Looking for raw/group repository: uan-2.5-sle-15sp3... Found...Updating raw/group repository: uan-2.5-sle-15sp3...204 OK updated
Looking for raw/hosted repository: uan-2.5.9-third-party... Found...Updating raw/hosted repository: uan-2.5.9-third-party...204 OK updated
Looking for raw/group repository: uan-2.5-third-party... Found...Updating raw/group repository: uan-2.5-third-party...204 OK updated
+ export SKOPEO_IMAGE=docker.io/library/skopeo:uan-2.5.9-7-g32bfb7f
+ SKOPEO_IMAGE=docker.io/library/skopeo:uan-2.5.9-7-g32bfb7f
+ export CRAY_NEXUS_SETUP_IMAGE=docker.io/library/cray-nexus-setup:uan-2.5.9-7-g32bfb7f
+ CRAY_NEXUS_SETUP_IMAGE=docker.io/library/cray-nexus-setup:uan-2.5.9-7-g32bfb7f
+ [[ -f ./lib/nexus-upload.sh ]]
+ . ./lib/nexus-upload.sh
+++ dirname ./lib/nexus-upload.sh
++ ROOTDIR=./lib/..
++ source ./lib/../lib/install.sh
+++ : https://packages.local
+++ : registry.local
+++ export NEXUS_URL
+++ : ./lib/..
+++ podman_run_flags=(--network host)
+++ declare -a podman_run_flags
+++ [[ no == \y\e\s ]]
+++ requires curl find podman realpath
+++ [[ 4 -gt 0 ]]
+++ command -v curl
+++ shift
+++ [[ 3 -gt 0 ]]
+++ command -v find
+++ shift
+++ [[ 2 -gt 0 ]]
+++ command -v podman
+++ shift
+++ [[ 1 -gt 0 ]]
+++ command -v realpath
+++ shift
+++ [[ 0 -gt 0 ]]
+++ vendor_images=()
+++ declare -a vendor_images
++ skopeo-sync ./lib/../docker
++ local src=./lib/../docker
++ [[ -d ./lib/../docker ]]
++ nexus-setdefault-credential
++ [[ -v NEXUS_PASSWORD ]]
++ [[ -n EcvveawHHqtQPkAOEapCdRiVDYUrmz5VIr9FE6-vanU= ]]
++ return 0
+++ realpath ./lib/../docker
++ podman run --rm --network host -v /mnt/developer/alanm/uan-2.5.9-7-g32bfb7f/docker:/image:ro docker.io/library/skopeo:uan-2.5.9-7-g32bfb7f sync --scoped --src dir --dest docker --dest-creds admin:EcvveawHHqtQPkAOEapCdRiVDYUrmz5VIr9FE6-vanU= --dest-tls-verify=false /image registry.local
time="2023-03-22T21:07:36Z" level=info msg="Copying image ref 1/9" from="dir:/image/artifactory.algol60.net/csm-docker/stable/cray-product-catalog-update:1.3.1" to="docker://registry.local/artifactory.algol60.net/csm-docker/stable/cray-product-catalog-update:1.3.1"
Getting image source signatures
Skipping: image already present at destination
time="2023-03-22T21:07:36Z" level=info msg="Copying image ref 2/9" from="dir:/image/artifactory.algol60.net/csm-docker/stable/docker.io/haproxytech/haproxy-alpine:2.6.6" to="docker://registry.local/artifactory.algol60.net/csm-docker/stable/docker.io/haproxytech/haproxy-alpine:2.6.6"
Getting image source signatures
Skipping: image already present at destination
time="2023-03-22T21:07:36Z" level=info msg="Copying image ref 3/9" from="dir:/image/artifactory.algol60.net/csm-docker/stable/docker.io/rancher/local-path-provisioner:v0.0.23" to="docker://registry.local/artifactory.algol60.net/csm-docker/stable/docker.io/rancher/local-path-provisioner:v0.0.23"
Getting image source signatures
Copying blob sha256:213ec9aee27d8be045c6a92b7eac22c9a64b44558193775a1a7f626352392b49
Copying blob sha256:d6dfa94d1f44ba06c355fb5a651ae27dbc0881114f03305e1ea7cc52788b2847
Copying blob sha256:b55e83f10d4df03312f9d081230e190d91d51573ee8cb116a48347925b8a08e4
Copying blob sha256:bec39189e70357237cb6a9295b944b423ee3314cce206ce116709b04c96ceb54
Copying blob sha256:5513eda6c57cbef7402659fc38acefac97689618a516c5f04004f5c6b4c09819
Copying config sha256:ff6a2695e5da67422e107193c07797cfce832b1413a1c44ae8d24c2602321cd8
Writing manifest to image destination
Storing signatures
time="2023-03-22T21:07:37Z" level=info msg="Copying image ref 4/9" from="dir:/image/artifactory.algol60.net/csm-docker/stable/docker.io/rancher/mirrored-coredns-coredns:1.9.4" to="docker://registry.local/artifactory.algol60.net/csm-docker/stable/docker.io/rancher/mirrored-coredns-coredns:1.9.4"
Getting image source signatures
Copying blob sha256:c6824c7a0594e5640ccb153e26696a9f13919ea95bad5c91e43ae298868af828
Copying blob sha256:8f16f0bc6a9b0a6c4f176312d4edb4d2ce9bf05be1321a681f876733a228e9a9
Copying config sha256:9a4fad1223994358079792536f8e66f35cf8dff8fd5bd3b2ece6197297641d09
Writing manifest to image destination
Storing signatures
time="2023-03-22T21:07:38Z" level=info msg="Copying image ref 5/9" from="dir:/image/artifactory.algol60.net/csm-docker/stable/docker.io/rancher/mirrored-metrics-server:v0.6.2" to="docker://registry.local/artifactory.algol60.net/csm-docker/stable/docker.io/rancher/mirrored-metrics-server:v0.6.2"
Getting image source signatures
Copying blob sha256:8fdb1fc20e240e9cae976518305db9f9486caa155fd5fc53e7b3a3285fe8a990
Copying blob sha256:81c08c324550fd5d554064975aed9e5754d767a7ffa4557617edf4ce3caf95d7
Copying config sha256:bba68dfd5f3968afa50d83bb80f94b4153199b3e29e823f08c691fdde02263a6
Writing manifest to image destination
Storing signatures
time="2023-03-22T21:07:40Z" level=info msg="Copying image ref 6/9" from="dir:/image/artifactory.algol60.net/csm-docker/stable/docker.io/rancher/mirrored-pause:3.6" to="docker://registry.local/artifactory.algol60.net/csm-docker/stable/docker.io/rancher/mirrored-pause:3.6"
Getting image source signatures
Copying blob sha256:fbe1a72f5dcd08ba4ca3ce3468c742786c1f6578c1f6bb401be1c4620d6ff705
Copying config sha256:12242b79542044ffc7daf2c6db797c8f5d65ac9ffa3fd5c8355d3ea1ec23a122
Writing manifest to image destination
Storing signatures
time="2023-03-22T21:07:40Z" level=info msg="Copying image ref 7/9" from="dir:/image/artifactory.algol60.net/csm-docker/stable/quay.io/metallb/controller:v0.13.7" to="docker://registry.local/artifactory.algol60.net/csm-docker/stable/quay.io/metallb/controller:v0.13.7"
Getting image source signatures
Skipping: image already present at destination
time="2023-03-22T21:07:40Z" level=info msg="Copying image ref 8/9" from="dir:/image/artifactory.algol60.net/csm-docker/stable/quay.io/metallb/speaker:v0.13.7" to="docker://registry.local/artifactory.algol60.net/csm-docker/stable/quay.io/metallb/speaker:v0.13.7"
Getting image source signatures
Skipping: image already present at destination
time="2023-03-22T21:07:40Z" level=info msg="Copying image ref 9/9" from="dir:/image/artifactory.algol60.net/uan-docker/stable/cray-uan-config:1.9.11" to="docker://registry.local/artifactory.algol60.net/uan-docker/stable/cray-uan-config:1.9.11"
Getting image source signatures
Skipping: image already present at destination
time="2023-03-22T21:07:40Z" level=info msg="Synced 9 images from 1 sources"
++ nexus-upload helm ./lib/../helm charts
++ local repotype=helm
++ local src=./lib/../helm
++ local reponame=charts
++ [[ -d ./lib/../helm ]]
++ nexus-setdefault-credential
++ [[ -v NEXUS_PASSWORD ]]
++ [[ -n EcvveawHHqtQPkAOEapCdRiVDYUrmz5VIr9FE6-vanU= ]]
++ return 0
+++ realpath ./lib/../helm
++ podman run --rm --network host -v /mnt/developer/alanm/uan-2.5.9-7-g32bfb7f/helm:/data:ro -e NEXUS_URL -e NEXUS_USERNAME -e NEXUS_PASSWORD docker.io/library/cray-nexus-setup:uan-2.5.9-7-g32bfb7f nexus-upload-repo-helm /data/ charts
/data/cray-uan-install-1.9.11.tgz	400 3581 0.072312 49521
/data/cray-uan-install-1.9.11.tgz	exists
real	0m 0.40s
user	0m 0.24s
sys	0m 0.11s
++ nexus-upload raw ./lib/../rpms/sle-15sp4 uan-2.5.9-sle-15sp4
++ local repotype=raw
++ local src=./lib/../rpms/sle-15sp4
++ local reponame=uan-2.5.9-sle-15sp4
++ [[ -d ./lib/../rpms/sle-15sp4 ]]
++ nexus-setdefault-credential
++ [[ -v NEXUS_PASSWORD ]]
++ [[ -n EcvveawHHqtQPkAOEapCdRiVDYUrmz5VIr9FE6-vanU= ]]
++ return 0
+++ realpath ./lib/../rpms/sle-15sp4
++ podman run --rm --network host -v /mnt/developer/alanm/uan-2.5.9-7-g32bfb7f/rpms/sle-15sp4:/data:ro -e NEXUS_URL -e NEXUS_USERNAME -e NEXUS_PASSWORD docker.io/library/cray-nexus-setup:uan-2.5.9-7-g32bfb7f nexus-upload-repo-raw /data/ uan-2.5.9-sle-15sp4
/data/noarch/cray-uan-systemd-presets-1.2.0-1.noarch.rpm	400 7621 0.038895 195937
/data/noarch/cray-uan-systemd-presets-1.2.0-1.noarch.rpm	exists
/data/repodata/cdec88780cbf36b550086e9b2119546af34a1faaf304875af23b152b4eaca47e-filelists.xml.gz	400 1360 0.059086 23017
/data/repodata/cdec88780cbf36b550086e9b2119546af34a1faaf304875af23b152b4eaca47e-filelists.xml.gz	exists
/data/x86_64/cray-uan-load-drivers-shasta-uan-1.2.0-1.x86_64.rpm	400 7397 0.066360 111467
/data/x86_64/cray-uan-load-drivers-shasta-uan-1.2.0-1.x86_64.rpm	exists
/data/repodata/a65e608037340dd786586887050283e9cf824f9d0ed9ee6bfe96c9c4577b488a-primary.sqlite.bz2	400 3864 0.074712 51718
/data/repodata/a65e608037340dd786586887050283e9cf824f9d0ed9ee6bfe96c9c4577b488a-primary.sqlite.bz2	exists
/data/x86_64/cray-uan-diagnostics-1.2.0-1.x86_64.rpm	400 12957 0.068785 188369
/data/x86_64/cray-uan-diagnostics-1.2.0-1.x86_64.rpm	exists
/data/repodata/5a70340b5e19d34365eb028f969681bf5268ef42bfc6acb7ee702a9242a87f8b-primary.xml.gz	400 2123 0.088856 23892
/data/repodata/5a70340b5e19d34365eb028f969681bf5268ef42bfc6acb7ee702a9242a87f8b-primary.xml.gz	exists
/data/x86_64/cray-uan-boot-parameters-shasta-uan-1.2.0-1.x86_64.rpm	400 7667 0.076947 99640
/data/x86_64/cray-uan-boot-parameters-shasta-uan-1.2.0-1.x86_64.rpm	exists
/data/repodata/repomd.xml	400 3501 0.087604 39963
/data/repodata/repomd.xml	exists
/data/repodata/76a81592f7c2a15c1dcdad1e3420d7448f1e07e7d856921cb6adc5fad4d64a33-other.xml.gz	400 1050 0.084283 12458
/data/repodata/76a81592f7c2a15c1dcdad1e3420d7448f1e07e7d856921cb6adc5fad4d64a33-other.xml.gz	exists
/data/repodata/b977c733d4359a9b1d4ae0d800c9b708a1c839c0e5a40b73f23e6fd76482d9d3-other.sqlite.bz2	400 1544 0.094820 16283
/data/repodata/b977c733d4359a9b1d4ae0d800c9b708a1c839c0e5a40b73f23e6fd76482d9d3-other.sqlite.bz2	exists
/data/repodata/625bbb65ad653e1f9547eaa5364ebecf7577813f7c8b934e6bf6fa4b683c8b40-filelists.sqlite.bz2	400 2252 0.090500 24883
/data/repodata/625bbb65ad653e1f9547eaa5364ebecf7577813f7c8b934e6bf6fa4b683c8b40-filelists.sqlite.bz2	exists
/data/noarch/cray-uan-goss-0.3.18-1.noarch.rpm	400 3280457 0.119777 27388037
/data/noarch/cray-uan-goss-0.3.18-1.noarch.rpm	exists
real	0m 0.43s
user	0m 0.41s
sys	0m 0.22s
++ nexus-upload raw ./lib/../rpms/sle-15sp3 uan-2.5.9-sle-15sp3
++ local repotype=raw
++ local src=./lib/../rpms/sle-15sp3
++ local reponame=uan-2.5.9-sle-15sp3
++ [[ -d ./lib/../rpms/sle-15sp3 ]]
++ nexus-setdefault-credential
++ [[ -v NEXUS_PASSWORD ]]
++ [[ -n EcvveawHHqtQPkAOEapCdRiVDYUrmz5VIr9FE6-vanU= ]]
++ return 0
+++ realpath ./lib/../rpms/sle-15sp3
++ podman run --rm --network host -v /mnt/developer/alanm/uan-2.5.9-7-g32bfb7f/rpms/sle-15sp3:/data:ro -e NEXUS_URL -e NEXUS_USERNAME -e NEXUS_PASSWORD docker.io/library/cray-nexus-setup:uan-2.5.9-7-g32bfb7f nexus-upload-repo-raw /data/ uan-2.5.9-sle-15sp3
/data/noarch/cray-uan-systemd-presets-1.2.0-1.noarch.rpm	400 7545 0.048032 157082
/data/noarch/cray-uan-systemd-presets-1.2.0-1.noarch.rpm	exists
/data/repodata/cdefae8ed27c864143b2936215bb9a77c6b064897d9c3801ba9a23e172ba2bb2-primary.xml.gz	400 2080 0.060660 34289
/data/repodata/cdefae8ed27c864143b2936215bb9a77c6b064897d9c3801ba9a23e172ba2bb2-primary.xml.gz	exists
/data/repodata/4c5bbce9673ecdbd4ae326c5f3c573e2ada75df9beff66a9a3722ec74452334d-filelists.sqlite.bz2	400 2255 0.070944 31785
/data/repodata/4c5bbce9673ecdbd4ae326c5f3c573e2ada75df9beff66a9a3722ec74452334d-filelists.sqlite.bz2	exists
/data/repodata/9908e851c013815f2d28e9640588419c892851a41e843fb12263f755140d3f73-filelists.xml.gz	400 1357 0.066722 20338
/data/repodata/9908e851c013815f2d28e9640588419c892851a41e843fb12263f755140d3f73-filelists.xml.gz	exists
/data/repodata/repomd.xml	400 3501 0.063980 54720
/data/repodata/repomd.xml	exists
/data/repodata/a0893706bbf853f62559bd32bd0574c42964a37e9f418be57e35f7523d897e1a-other.xml.gz	400 1048 0.061826 16950
/data/repodata/a0893706bbf853f62559bd32bd0574c42964a37e9f418be57e35f7523d897e1a-other.xml.gz	exists
/data/repodata/dbcf902183d65e6e1c45e34c2b50d401fd050c7d34d30365c393ae925a4f76e7-primary.sqlite.bz2	400 3793 0.063139 60073
/data/repodata/dbcf902183d65e6e1c45e34c2b50d401fd050c7d34d30365c393ae925a4f76e7-primary.sqlite.bz2	exists
/data/x86_64/cray-uan-boot-parameters-shasta-uan-1.2.0-1.x86_64.rpm	400 7539 0.060251 125126
/data/x86_64/cray-uan-boot-parameters-shasta-uan-1.2.0-1.x86_64.rpm	exists
/data/x86_64/cray-uan-diagnostics-1.2.0-1.x86_64.rpm	400 12829 0.058894 217832
/data/x86_64/cray-uan-diagnostics-1.2.0-1.x86_64.rpm	exists
/data/repodata/6bcef5b972de60034437624225afa798f08c69e655176ae56d4e918fa81566ea-other.sqlite.bz2	400 1529 0.069310 22060
/data/repodata/6bcef5b972de60034437624225afa798f08c69e655176ae56d4e918fa81566ea-other.sqlite.bz2	exists
/data/x86_64/cray-uan-load-drivers-shasta-uan-1.2.0-1.x86_64.rpm	400 7269 0.072170 100720
/data/x86_64/cray-uan-load-drivers-shasta-uan-1.2.0-1.x86_64.rpm	exists
/data/noarch/cray-uan-goss-0.3.18-1.noarch.rpm	400 3279625 0.112921 29043534
/data/noarch/cray-uan-goss-0.3.18-1.noarch.rpm	exists
real	0m 0.41s
user	0m 0.40s
sys	0m 0.20s
++ nexus-upload raw ./lib/../third-party uan-2.5.9-third-party
++ local repotype=raw
++ local src=./lib/../third-party
++ local reponame=uan-2.5.9-third-party
++ [[ -d ./lib/../third-party ]]
++ nexus-setdefault-credential
++ [[ -v NEXUS_PASSWORD ]]
++ [[ -n EcvveawHHqtQPkAOEapCdRiVDYUrmz5VIr9FE6-vanU= ]]
++ return 0
+++ realpath ./lib/../third-party
++ podman run --rm --network host -v /mnt/developer/alanm/uan-2.5.9-7-g32bfb7f/third-party:/data:ro -e NEXUS_URL -e NEXUS_USERNAME -e NEXUS_PASSWORD docker.io/library/cray-nexus-setup:uan-2.5.9-7-g32bfb7f nexus-upload-repo-raw /data/ uan-2.5.9-third-party
/data/k3s-airgap-images-amd64.tar.md5sum	400 510 0.027250 18715
/data/k3s-airgap-images-amd64.tar.md5sum	exists
/data/k3s-install.sh.md5sum	400 484 0.050210 9639
/data/k3s-install.sh.md5sum	exists
/data/haproxy-1.17.3.tgz	400 30991 0.049456 626637
/data/haproxy-1.17.3.tgz	exists
/data/metallb-0.13.7.tgz	400 23372 0.049137 475649
/data/metallb-0.13.7.tgz	exists
/data/k3s.md5sum	400 462 0.048042 9616
/data/k3s.md5sum	exists
/data/k3s-install.sh	400 28289 0.052976 533996
/data/k3s-install.sh	exists
/data/k3s	400 65986975 1.560306 42291047
/data/k3s	exists
/data/k3s-airgap-images-amd64.tar	400 554935247 12.736004 43572163
/data/k3s-airgap-images-amd64.tar	exists
real	0m 13.07s
user	0m 0.72s
sys	0m 0.91s
+ clean-install-deps
+ crictl info
+ jq -e '.config.registry.mirrors | keys | map(select(. == "artifactory.algol60.net")) | length <= 0'
+ loftsman ship --charts-path ./lib/../helm --manifest-path ./lib/../build/manifests/uan.yaml
2023-03-22T21:07:57Z INF Initializing the connection to the Kubernetes cluster using KUBECONFIG (system default), and context (current-context) command=ship
2023-03-22T21:07:57Z INF Initializing helm client object command=ship
         |\
         | \
         |  \
         |___\      Shipping your Helm workloads with Loftsman
       \--||___/
  ~~~~~~\_____/~~~~~~~

2023-03-22T21:07:57Z INF Ensuring that the loftsman namespace exists command=ship
2023-03-22T21:07:57Z INF Loftsman will use the packaged charts at ./lib/../helm as the Helm install source command=ship
2023-03-22T21:07:57Z INF Running a release for the provided manifest at ./lib/../build/manifests/uan.yaml command=ship

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Releasing cray-uan-install v1.9.11
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

2023-03-22T21:07:57Z INF Found value overrides for chart, applying:
cray-import-config:
  import_job:
    CF_IMPORT_PRODUCT_VERSION: 2.5.9-7-g32bfb7f
 chart=cray-uan-install command=ship namespace=services version=1.9.11
2023-03-22T21:07:57Z INF Running helm install/upgrade with arguments: upgrade --install cray-uan-install helm/cray-uan-install-1.9.11.tgz --namespace services --create-namespace --set global.chart.name=cray-uan-install --set global.chart.version=1.9.11 -f /tmp/loftsman-1679519277/cray-uan-install-values.yaml chart=cray-uan-install command=ship namespace=services version=1.9.11
2023-03-22T21:07:58Z INF Release "cray-uan-install" has been upgraded. Happy Helming!
NAME: cray-uan-install
LAST DEPLOYED: Wed Mar 22 21:07:58 2023
NAMESPACE: services
STATUS: deployed
REVISION: 11
TEST SUITE: None
 chart=cray-uan-install command=ship namespace=services version=1.9.11
2023-03-22T21:07:58Z INF Ship status: success. Recording status, manifest to configmap loftsman-uan in namespace loftsman command=ship
2023-03-22T21:07:58Z INF Recording log data to configmap loftsman-uan-ship-log in namespace loftsman command=ship
+ ARTIFACT_PATH=./lib/../images/application
+ KERNEL=./lib/../images/application/5.3.18-150300.59.87-default.kernel
+ INITRD=./lib/../images/application/initrd.img.xz
+ ROOTFS=./lib/../images/application/application.squashfs
++ list_ims_images
++ set +x
++ jq --arg UAN_IMAGE_NAME cray-application-sles15sp3.x86_64-0.1.0 -r 'sort_by(.created) | .[] | select(.name == $UAN_IMAGE_NAME ) | .id'
++ head -1
+ IMAGE_ID=63132c53-ba5e-4b08-bbc8-3c1571d8f3ee
+ '[' -z 63132c53-ba5e-4b08-bbc8-3c1571d8f3ee ']'
+ echo 'Found cray-application-sles15sp3.x86_64-0.1.0 already exists as 63132c53-ba5e-4b08-bbc8-3c1571d8f3ee... Skipping image upload'
Found cray-application-sles15sp3.x86_64-0.1.0 already exists as 63132c53-ba5e-4b08-bbc8-3c1571d8f3ee... Skipping image upload
+ '[' -z 63132c53-ba5e-4b08-bbc8-3c1571d8f3ee ']'
+ cat
+ podman run --rm --name uan-2.5.9-7-g32bfb7f-image-catalog-update -u root -e PRODUCT=uan -e PRODUCT_VERSION=2.5.9-7-g32bfb7f -e YAML_CONTENT=/results/2.5.9-7-g32bfb7f-63132c53-ba5e-4b08-bbc8-3c1571d8f3ee.json -e KUBECONFIG=/.kube/admin.conf -v /etc/kubernetes:/.kube:ro -v /mnt/developer/alanm/uan-2.5.9-7-g32bfb7f:/results:ro registry.local/artifactory.algol60.net/csm-docker/stable/cray-product-catalog-update:1.3.1
Updating config_map=cray-product-catalog in namespace=services for product/version=uan/2.5.9-7-g32bfb7f
Retrieving content from /results/2.5.9-7-g32bfb7f-63132c53-ba5e-4b08-bbc8-3c1571d8f3ee.json
Resting 3s before reading ConfigMap
Version=2.5.9-7-g32bfb7f does not exist; will update
ConfigMap update attempt=1
Resting 2s before reading ConfigMap
ConfigMap data updates exist; Exiting.
```

alanm-hpe commented:
UAN Bootprep file used to test on groot
```
ncn-m001:/mnt/developer/alanm/bootprep # cat alanm-bootprep.yaml
# (C) Copyright 2022 Hewlett Packard Enterprise Development LP
---
configurations:
- name: alanm-uan-config-cos-{{cos.version}}
  layers:
  - name: shs-mellanox_install-integration-{{shs.version}}
    playbook: shs_mellanox_install.yml
    product:
      name: slingshot-host-software
      version: "{{shs.version}}"
      branch: integration-{{shs.version}}
  - name: csm-packages-integration-{{csm.version}}
    playbook: csm_packages.yml
    product:
      name: csm
      version: "{{csm.version}}"
  - name: uan
    playbook: site.yml
    product:
      name: uan
      version: "{{uan.version}}"
      branch: integration-uan-2.5.9-6-ge2e5882
  - name: uan
    playbook: k3s.yml
    product:
      name: uan
      version: "{{uan.version}}"
      branch: integration-uan-2.5.9-6-ge2e5882
  - name: cpe-pe_deploy-integration-{{cpe.version}}
    playbook: pe_deploy.yml
    product:
      name: cpe
      version: "{{cpe.version}}"
      branch: cpe-{{cpe.version.split('.')[0]}}.{{cpe.version.split('.')[1].zfill(2)}}-integration
  - name: analytics-site-integration-{{analytics.version}}
    playbook: site.yml
    product:
      name: analytics
      version: "{{analytics.version}}"
      branch: integration
  - name: slurm-site-{{slurm.version}}
    playbook: site.yml
    product:
      name: slurm
      version: "{{slurm.version}}"
      commit: "53bfbf99074f607336d11b7669f4e39171c5f516"

images:
- name: alanm-uan-kpreslan
  ims:
    is_recipe: false
    #name: kpreslan-20230313190241_cfs_kpreslan-20230313190241
    id: 28c4ab84-7ec1-42eb-a431-7ee0bdbf3060
  configuration: alanm-uan-config-cos-{{cos.version}}
  configuration_group_names:
  - Application
  - Application_UAN

session_templates:
- name: alanm-uan-all-cos-{{cos.version}}
  image: alanm-uan-kpreslan
  configuration: alanm-uan-config-cos-{{cos.version}}
  bos_parameters:
    boot_sets:
      uan:
        kernel_parameters: spire_join_token=${SPIRE_JOIN_TOKEN}
        node_roles_groups:
        - Application
        # If BOS v2 will be used to create the session from this session template,
        # you can target the UAN subrole here instead. E.g.:
        #
        # node_roles_groups:
        # - Application_UAN
        #
        # If not using BOS v2, and the system has other nodes with the
        # "Application" role in HSM that are not UANs, use node_list instead of
        # node_roles_groups. E.g.:
        #
        # node_list:
        # - xname1
        # - xname2
        rootfs_provider_passthrough: "dvs:api-gw-service-nmn.local:300:hsn0,nmn0:0"
```

Product vars tested:
```
ncn-m001:/mnt/developer/alanm/bootprep # cat product_vars.yaml
# Copyright 2022 Hewlett Packard Enterprise Development LP
---
analytics:
  version: 1.3.18
cos:
  version: 2.4.109
cpe:
  version: 22.12.5
csm:
  version: 1.3.1
csm_diags:
  version: 1.3.28
hfp:
  version: 23.01.0
pbs:
  version: 1.2.3
#recipe:
#  version: 22.11.0-alpha.16
sat:
  version: 2.4.13
sdu:
  version: 2.1.3
shs:
  version: 2.0.2-97
sle_sp3:
  version: 22.08.0.20220805
sle_sp4:
  version: 23.02.0.20230216
slingshot:
  version: 2.0.2-941
slurm:
  version: 1.2.8
sma:
  version: 1.7.27
uan:
  version: 2.5.9-7-g32bfb7f
```

Note, a custom image was used due to MQTT DVS testing. It is expected that the image used will be representative of the image built with COS 2.4.111
